### ISSUE TITLE
Can't load can-j1939 module [Segmentation fault]

### ISSUE THREAD

**Issue by maleselo** [kurt-vd/linux#2]
Hi,

I'm trying to use the can-j1939 module but I can't load the can-j1939 module, I have got a segmentation fault within the j1939_netdev_notify function.

The stack trace is here :
```
/ # modprobe vcan
[   11.623243] vcan: Virtual CAN interface driver
/ # ip link add vcan0 type vcan
[   20.666573] ip (971) used greatest stack depth: 13312 bytes left
/ # ip link set vcan0 up
/ # modprobe can-j1939
[   39.642954] can: SAE J1939
[   39.643545] BUG: stack guard page was hit at ffffc900000d81a8 (stack is ffffc900000d4000..ffffc900000d7fff)
[   39.644192] kernel stack overflow (page fault): 0000 [#1] SMP
[   39.644192] Modules linked in: can_j1939(+) vcan
[   39.644192] CPU: 0 PID: 980 Comm: modprobe Not tainted 4.9.68+ #5
[   39.644192] Hardware name: QEMU Standard PC (i440FX + PIIX, 1996), BIOS Ubuntu-1.8.2-1ubuntu1 04/01/2014
[   39.644192] task: ffff8800062a6c00 task.stack: ffffc900000d4000
[   39.644192] RIP: 0010:[<ffffffffa0005160>]  [<ffffffffa0005160>] j1939_netdev_notify+0x0/0x60 [can_j1939]
[   39.644192] RSP: 0018:ffffc900000d7c98  EFLAGS: 00000206
[   39.644192] RAX: ffff8800071d0050 RBX: ffffffffa0009ca0 RCX: ffff880007381dc1
[   39.644192] RDX: ffffc900000d7ca8 RSI: 0000000000000005 RDI: ffffffffa0009ca0
[   39.644192] RBP: ffffc900000d7cd8 R08: ffff880007381dc0 R09: ffff8800072d8040
[   39.644192] R10: ffff880007003d00 R11: 000000000000015e R12: ffffffffa000a0c0
[   39.644192] R13: ffffffff81ee7e98 R14: ffff8800071d0000 R15: ffffffff81ee7d80
[   39.644192] FS:  0000000000ce9880(0000) GS:ffff880007c00000(0000) knlGS:0000000000000000
[   39.644192] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[   39.644192] CR2: ffffc900000d81a8 CR3: 000000000630b000 CR4: 00000000000006f0
[   39.644192] Stack:
[   39.644192]  ffffffff8171db25 ffffc90000000000 ffff8800071d0000 0000000000000000
[   39.644192]  ffffffffa000a0c0 ffffffffa000d000 ffffc900000d7eb0 ffff8800071f4c00
[   39.644192]  ffffc900000d7cf0 ffffffffa000d045 0000000000000000 ffffc900000d7d68
[   39.644192] Call Trace:
[   39.644192]  [<ffffffff8171db25>] ? register_netdevice_notifier+0x85/0x1c0
[   39.644192]  [<ffffffffa000d000>] ? 0xffffffffa000d000
[   39.644192]  [<ffffffffa000d045>] j1939_module_init+0x45/0x10e [can_j1939]
[   39.644192]  [<ffffffff81000408>] do_one_initcall+0x38/0x140
[   39.644192]  [<ffffffff81128d8a>] ? do_init_module+0x22/0x1c8
[   39.644192]  [<ffffffff81908936>] ? _cond_resched+0x26/0x40
[   39.644192]  [<ffffffff8117be63>] ? kmem_cache_alloc_trace+0x33/0x150
[   39.644192]  [<ffffffff81128dc2>] do_init_module+0x5a/0x1c8
[   39.644192]  [<ffffffff810d0fd8>] load_module+0x1e78/0x2340
[   39.644192]  [<ffffffff810ce2e0>] ? __symbol_put+0x40/0x40
[   39.644192]  [<ffffffff81180001>] ? do_pages_stat+0x71/0x170
[   39.644192]  [<ffffffff8118b5c4>] ? kernel_read_file_from_fd+0x44/0x70
[   39.644192]  [<ffffffff810d168a>] SYSC_finit_module+0xba/0xc0
[   39.644192]  [<ffffffff810d16a9>] SyS_finit_module+0x9/0x10
[   39.644192]  [<ffffffff8190caa0>] entry_SYSCALL_64_fastpath+0x13/0x94
[   39.644192] Code: ff ff ff be 2e 00 00 00 48 c7 c7 91 99 00 a0 c6 05 d7 4d 00 00 01 e8 40 4c 05 e1 e9 5b ff ff ff 90 66 2e 0f 1f 84 00 00 00 00 00 <48> 81 ba 00 05 00 00 80 7d ee 81 74 03 31 c0 c3 66 81 ba 34 02 
[   39.644192] RIP  [<ffffffffa0005160>] j1939_netdev_notify+0x0/0x60 [can_j1939]
[   39.644192]  RSP <ffffc900000d7c98>
[   39.644192] ---[ end trace 8725fff2a94188ea ]---
Segmentation fault

```
Is there any prerequisite I should set during the kernel configuration to fix this ?

Regards

**Comment by kurt-vd** [comment]
Hey,

>    Hi,
> 
>    I'm trying to use the can-j1939 module but I can't load the can-j1939
>    module, I have got a segmentation fault within the j1939_netdev_notify
>    function.

Well, this supprises me a bit.
Are you sure you run a consistent kernel (i.e. all modules, including
can-j1939) are compiled from the same source & configuration?

I have no immediate anser to what can cause a stack overflow.
Maybe post this to linux-can@vger.kernel.org?

Kind regards,
Kurt

**Comment by maleselo** [comment]
Hi,

I did the procedure described here: https://elinux.org/J1939. 

I downloaded a 4.9.68 linux kernel, I merged it with the j1939d-v4.7 branch of this repository and I enabled SAE J1939 module in the kernel configuration.

I made the kernel and the modules and then I installed the modules in a filesystem created via busybox. I emulated my distribution with Qemu. 

I can load the vcan module but when it comes to the can-j1939 module i get this error above.

Do you have any examples (kernel configs ...), ressources I could use to make it work ? Should I use a particular kernel version ? 

Regards

**EDIT: I misunderstood the procedure, I merged the j1939d-v4.9.10 branch with a 4.7.1 linux kernel. I tried with the j1939d-v4.7 branch and now I'm able to load the SAE J1939 module correctly.**

**Comment by kurt-vd** [comment]
Sorry for the late asnwer.
>    Hi,
>    I did the procedure described here: [1]https://elinux.org/J1939.
> 
>    I downloaded a 4.9.68 linux kernel, I merged it with j1939d-v4.7 branch
>    of this repository and I enabled SAE J1939 module in the kernel
>    configuration.
> 
>    I installed the module in a filesystem created via busybox and then I
>    emulate my kernel + my fs with Qemu.
> 
>    I can load the vcan module but when it comes to the can-j1939 module i
>    get this error above.
> 
>    Do you have any examples (kernel configs ...) I could use to make it
>    work ? Should I use a particular kernel version ?

Can you try my can-next-j1939 branch? It has been tested way better
recently.

In general, segmentation faults like this, while just loading the
modules, are mostly caused by loading modules that were built using a
different kernel source or config, so maybe try just running a kernel
with everything built in (i.e. without modules), that way you can
eliminate build problems.

Kind regards,
Kurt

**Comment by maleselo** [comment]
Hi,

Yes it is working. My previous problem was caused by a wrong kernel version indeed.

Many thanks,
Maleselo
